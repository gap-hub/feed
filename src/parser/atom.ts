import * as convert from "xml-js";

import { defaultTextType } from "../config";
import { Feed } from "../feed";
import { FeedItem } from "../feed-item";

export function parseAtom(xml: convert.ElementCompact): Feed {
  const feedXml = xml.feed;
  const feed = new Feed({
    id: feedXml.id?._text ?? "",
    title: feedXml.title?._text ?? "",
    updated: feedXml.updated?._text
      ? new Date(feedXml.updated?._text)
      : new Date(),
    copyright: feedXml.rights?._text ?? "",
    generator: feedXml.generator?._text,
    description: feedXml.subtitle?._text,
    image: feedXml.logo?._text,
    favicon: feedXml.icon?._text,
  });

  if (feedXml.author) {
    feed.options.authors = [
      {
        name: feedXml.author.name?._text ?? "",
        email: feedXml.author.email?._text ?? "",
        link: feedXml.author.uri?._text ?? "",
      },
    ];
  }

  if (feedXml.link) {
    if (Array.isArray(feedXml.link)) {
      feedXml.link.forEach((item: any) => {
        buildFeedLink(item, feed);
      });
    } else {
      buildFeedLink(feedXml.link, feed);
    }
  }

  if (feedXml.category) {
    if (Array.isArray(feedXml.category)) {
      feedXml.category.forEach((item: any) => {
        if (item?._attributes?.term) {
          feed.addCategory(item._attributes.term);
        }
      });
    } else if (feedXml.category?._attributes?.term) {
      feed.addCategory(feedXml.category._attributes.term);
    }
  }

  if (feedXml.contributor) {
    if (Array.isArray(feedXml.contributor)) {
      feedXml.contributor.forEach((item: any) => {
        if (item.name?._text) {
          feed.addContributor({
            name: item.name?._text,
            email: item.email?._text,
            link: item.uri?._text,
          });
        }
      });
    } else if (feedXml.contributor.name?._text) {
      feed.addContributor({
        name: feedXml.contributor.name?._text,
        email: feedXml.contributor.email?._text,
        link: feedXml.contributor.uri?._text,
      });
    }
  }

  if (feedXml.entry) {
    if (Array.isArray(feedXml.entry)) {
      feedXml.entry.forEach((item: any) => {
        buildFeedItem(item, feed);
      });
    }
  } else {
    buildFeedItem(feedXml.entry, feed);
  }

  return feed;
}

function buildFeedItem(entry: convert.ElementCompact, feed: Feed) {
  const item = new FeedItem({
    id: entry.id?._text ?? "",
    title: {
      text: entry.title?._cdata ?? entry.title?._text ?? "",
      type: entry.title?._attributes?.type ?? defaultTextType,
    },
    link: entry.link?._attributes?.href ?? "",
    date: entry.updated?._text ? new Date(entry.updated?._text) : new Date(),
    description: {
      text: entry.summary?._cdata ?? entry.summary?._text ?? "",
      type: entry.summary?._attributes?.type ?? defaultTextType,
    },
    content: {
      text: entry.content?._cdata ?? entry.content?._text ?? "",
      type: entry.content?._attributes?.type ?? defaultTextType,
    },
    published: entry.published?._text
      ? new Date(entry.published?._text)
      : new Date(),
    copyright: entry.rights?._text,
  });

  if (entry.author) {
    if (Array.isArray(entry.author)) {
      entry.author.forEach((author: any) => {
        if (author.name?._text) {
          item.addAuthor({
            name: author.name?._text,
            email: author.email?._text,
            link: author.uri?._text,
          });
        }
      });
    } else if (entry.author.name?._text) {
      item.addAuthor({
        name: entry.author.name?._text,
        email: entry.author.email?._text,
        link: entry.author.uri?._text,
      });
    }
  }

  if (entry.category) {
    if (Array.isArray(entry.category)) {
      entry.category.forEach((category: any) => {
        if (category?._attributes?.label) {
          item.addCategory({
            name: category._attributes.label,
            term: category?._attributes?.term,
            scheme: category?._attributes?.scheme,
          });
        }
      });
    } else if (entry.category?._attributes?.label) {
      item.addCategory({
        name: entry.category._attributes.label,
        term: entry.category?._attributes?.term,
        scheme: entry.category?._attributes?.scheme,
      });
    }
  }

  if (entry.contributor) {
    if (Array.isArray(entry.contributor)) {
      entry.contributor.forEach((contributor: any) => {
        if (contributor.name?._text) {
          item.addContributor({
            name: contributor.name._text,
            email: contributor.email?._text,
            link: contributor.uri?._text,
          });
        }
      });
    } else if (entry.contributor.name?._text) {
      item.addContributor({
        name: entry.contributor.name._text,
        email: entry.contributor.email?._text,
        link: entry.contributor.uri?._text,
      });
    }
  }

  feed.addItem(item);
}

function buildFeedLink(link: convert.ElementCompact, feed: Feed) {
  if (link?._attributes?.href) {
    const href = link._attributes.href.toString();
    const rel = link?._attributes?.rel;
    if (rel === "self") {
      feed.options.feed = href;
      feed.options.feedLinks = {
        ...feed.options.feedLinks,
        atom: href,
      };
    } else if (rel === "hub") {
      feed.options.hub = href;
    } else if (rel === "alternate") {
      feed.options.link = href;
    }
  }
}
