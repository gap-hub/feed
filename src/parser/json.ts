import { defaultTextType } from "../config";
import { Feed } from "../feed";
import { FeedItem } from "../feed-item";
import {
  Extension,
  Text,
  combinedFeedFields,
  combinedFeedItemFields,
} from "../typings";

const fields = {
  feed: combinedFeedFields,
  item: combinedFeedItemFields,
};

export function parseJSON(json: any): Feed {
  const feed = new Feed({
    id: json.id ?? "",
    title: json.title ?? "",
    copyright: "",
    link: json.home_page_url,
    description: json.description ?? "",
    image: json.icon,
    favicon: json.favicon,
    generator: json.generator ?? "",
    feedLinks: json.feed_url ? { json: json.feed_url } : undefined,
    language: json.language,
  });
  feed.options.authors = [];
  if (json.author) {
    feed.options.authors.push({
      name: json.author.name ?? "",
      email: json.author.email,
      link: json.author.url,
      avatar: json.author.avatar,
    });
  }
  if (Array.isArray(json.authors)) {
    json.authors.forEach((author: any) => {
      feed.options.authors?.push({
        name: author.name ?? "",
        email: author.email,
        link: author.url,
        avatar: author.avatar,
      });
    });
  }
  const [extensions, otherFields] = parseOtherFields(json, true);
  feed.extensions = extensions;
  feed.setCustomFields(otherFields);

  if (Array.isArray(json.items)) {
    json.items.forEach((item: any) => {
      const feedItem = new FeedItem({
        id: item.id ?? "",
        title: { text: item.title ?? "", type: defaultTextType },
        date: item.date_modified ? new Date(item.date_modified) : new Date(),
        link: item.url ?? "",
        description: { text: item.summary ?? "", type: defaultTextType },
        image: item.image,
        published: item.date_published
          ? new Date(item.date_published)
          : undefined,
      });
      let content: Text | undefined;
      if (item.content_text) {
        content = { text: item.content_text, type: "text" };
      } else if (item.content_html) {
        content = { text: item.content_html, type: "html" };
      }
      if (content) {
        feedItem.setContent(content);
      }

      if (item.author) {
        feedItem.addAuthor({
          name: item.author.name ?? "",
          email: item.author.email,
          link: item.author.url,
          avatar: item.author.avatar,
        });
      }
      if (Array.isArray(item.authors)) {
        item.authors.forEach((author: any) => {
          feedItem.addAuthor({
            name: author.name ?? "",
            email: author.email,
            link: author.url,
            avatar: author.avatar,
          });
        });
      }

      if (Array.isArray(item.tags)) {
        item.tags.forEach((tag: string) => {
          feedItem.addCategory(tag);
        });
      }

      const [extensions, otherFields] = parseOtherFields(item, false);
      feedItem.setExtensions(extensions);
      feedItem.setCustomFields(otherFields);

      feed.addItem(feedItem);
    });
  }

  return feed;
}

function parseOtherFields(
  json: any,
  isFeed: boolean,
): [Extension[], Record<string, string | string[]>] {
  const extensions: Extension[] = [];
  const otherFields: Record<string, string | string[]> = {};
  const builtinKeys: Array<string> = fields[isFeed ? "feed" : "item"];
  Object.keys(json).forEach((key) => {
    const value = json[key];
    if (builtinKeys.indexOf(key) !== -1) {
      return;
    }
    if (key.startsWith("_")) {
      extensions.push({ name: key, objects: value });
    } else {
      otherFields[key] = value;
    }
  });

  return [extensions, otherFields];
}
