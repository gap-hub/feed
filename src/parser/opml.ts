import * as convert from "xml-js";

import { Opml } from "../opml";
import {
  OpmlBodyOptions,
  OpmlHeadOptions,
  OpmlOutlineOptions,
} from "../typings";
import { isNumber } from "../utils";

export function parseOPML(text: string): Opml {
  const result: convert.ElementCompact = convert.xml2js(text, {
    compact: true,
  });
  if (!result?.opml) {
    throw new Error("Invalid OPML");
  }
  return new Opml({
    version: result.opml._attributes?.version ?? "",
    head: result.opml.head ? parseHead(result.opml.head) : {},
    body: result.opml.body ? parseBody(result.opml.body) : { outlines: [] },
  });
}

function parseHead(xml: convert.ElementCompact): OpmlHeadOptions {
  const head: OpmlHeadOptions = {};
  const keys = Object.keys(xml);
  keys.forEach((key) => {
    const value: string | number | undefined =
      xml[key]?._text || xml[key]?._cdata;
    if (!value) {
      return;
    }
    if (["dateCreated", "dateModified"].includes(key)) {
      head[key] = new Date(value);
    } else if (
      [
        "vertScrollState",
        "windowTop",
        "windowLeft",
        "windowBottom",
        "windowRight",
      ].includes(key)
    ) {
      const num = Number(value);
      if (!isNaN(num)) {
        head[key] = num;
      }
    } else if (key === "expansionState") {
      if (isNumber(value)) {
        head[key] = [value];
      } else {
        head[key] = value
          .split(",")
          .map((num) => Number(num.trim()))
          .filter((num) => !isNaN(num));
      }
    } else {
      head[key] = value;
    }
  });
  return head;
}

function parseBody(xml: convert.ElementCompact): OpmlBodyOptions {
  const body: OpmlBodyOptions = { outlines: [] };
  let outlinesXml = xml.outline;
  if (!Array.isArray(outlinesXml)) {
    outlinesXml = [outlinesXml];
  }
  const outlines: OpmlOutlineOptions[] = [];
  parseOutlines(outlinesXml, outlines);
  body.outlines = outlines;
  return body;
}

function parseOutlines(
  xml: convert.ElementCompact[],
  outlines: OpmlOutlineOptions[],
) {
  for (let i = 0, len = xml.length; i < len; i++) {
    const outlineXml = xml[i];
    const outline = parseOutline(outlineXml);
    outlines.push(outline);
    if (outlineXml.outline) {
      let children = outlineXml.outline;
      if (!Array.isArray(children)) {
        children = [children];
      }
      outline.outlines = [];
      parseOutlines(children, outline.outlines);
    }
  }
}

function parseOutline(xml: convert.ElementCompact): OpmlOutlineOptions {
  const outline: OpmlOutlineOptions = {
    text: "",
  };
  if (xml._attributes) {
    const keys = Object.keys(xml._attributes);
    keys.forEach((key) => {
      const value = xml._attributes![key];
      if (!value) {
        return;
      }
      if (key === "isComment" || key === "isBreakpoint") {
        outline[key] = value === "true";
      } else if (key === "created") {
        outline.created = new Date(value);
      } else {
        (outline as any)[key] = value;
      }
    });
  }
  return outline;
}
