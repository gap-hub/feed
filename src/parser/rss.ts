import * as convert from "xml-js";

import { defaultTextType } from "../config";
import { Feed } from "../feed";
import { FeedItem } from "../feed-item";
import {
  Author,
  Enclosure,
  combinedFeedFields,
  combinedFeedItemFields,
} from "../typings";

export function parseRSS(xml: convert.ElementCompact): Feed {
  if (xml["rdf:RDF"]) {
    return buildRSS1(xml);
  } else if (xml.rss?._attributes?.version?.match(/0\.9/)) {
    return buildRSS09(xml);
  }
  return buildRSS2(xml);
}

function buildRSS09(xml: convert.ElementCompact): Feed {
  let channel = xml.rss.channel;
  if (Array.isArray(channel) && channel.length > 0) {
    channel = channel[0];
  }
  const items = channel.item;
  const feed = buildRSS(channel, items);
  const stylesheetHref = parseStylesheet(xml);
  if (stylesheetHref) {
    feed.stylesheet = stylesheetHref;
  }
  return feed;
}

function buildRSS1(xml: convert.ElementCompact): Feed {
  let channel = xml["rdf:RDF"].channel;
  if (Array.isArray(channel) && channel.length > 0) {
    channel = channel[0];
  }
  const items = channel.item;
  const feed = buildRSS(channel, items);
  const stylesheetHref = parseStylesheet(xml);
  if (stylesheetHref) {
    feed.stylesheet = stylesheetHref;
  }
  return feed;
}

function buildRSS2(xml: convert.ElementCompact): Feed {
  let channel = xml.rss.channel;
  if (Array.isArray(channel) && channel.length > 0) {
    channel = channel[0];
  }
  const items = channel.item;
  const feed = buildRSS(channel, items);
  const stylesheetHref = parseStylesheet(xml);
  if (stylesheetHref) {
    feed.stylesheet = stylesheetHref;
  }
  return feed;
}

function buildRSS(
  channel: convert.ElementCompact,
  items: convert.ElementCompact,
): Feed {
  const feed = new Feed({
    id: "",
    title: channel?.title?._text ?? "",
    link: channel?.link?._text ?? "",
    description: channel?.description?._text ?? "",
    updated: channel?.lastBuildDate?._text
      ? new Date(channel.lastBuildDate._text)
      : new Date(),
    docs:
      channel?.docs?._text ?? "https://validator.w3.org/feed/docs/rss2.html",
    generator: channel?.generator?._text,
    copyright: channel?.copyright?._text,
    language: channel?.language?._text,
  });
  const ignoreKeys = [
    "id",
    "title",
    "link",
    "description",
    "lastBuildDate",
    "docs",
    "generator",
    "copyright",
    "language",
    "item",
  ];
  Object.keys(channel).forEach((key) => {
    if (ignoreKeys.includes(key)) {
      return;
    }
    if (key === "ttl") {
      if (channel?.ttl?._text) {
        const ttl = parseInt(channel?.ttl?._text);
        if (!isNaN(ttl)) {
          feed.options.ttl = ttl;
        }
      }
    } else if (key === "image") {
      if (channel.image) {
        feed.options.image = channel.image.url?._text;
      }
    } else if (key === "category") {
      if (channel.category) {
        if (Array.isArray(channel.category)) {
          channel.category.forEach((item) => {
            if (item._text) {
              feed.addCategory(item._text);
            }
          });
        } else if (channel.category._text) {
          feed.addCategory(channel.category._text);
        }
      }
    } else if (key === "atom:link") {
      if (Array.isArray(channel["atom:link"])) {
        channel["atom:link"].forEach((item) => {
          buildAtomLink(item, feed);
        });
      } else if (channel["atom:link"]?._attributes?.href) {
        buildAtomLink(channel["atom:link"], feed);
      }
    } else if (!combinedFeedFields.includes(key)) {
      const value = channel[key];
      if (Array.isArray(value)) {
        feed.setCustomField(
          key,
          value.map((item) => item?._text),
        );
      } else {
        feed.setCustomField(key, value?._text);
      }
    }
  });

  if (items) {
    if (Array.isArray(items)) {
      items.forEach((item) => {
        buildItem(item, feed);
      });
    } else {
      buildItem(items, feed);
    }
  }

  return feed;
}

function buildAtomLink(atomLink: convert.ElementCompact, feed: Feed) {
  if (atomLink?._attributes?.href) {
    const href = atomLink._attributes.href.toString();
    const rel = atomLink._attributes?.rel;
    if (rel === "self") {
      feed.options.feed = href;
    } else if (rel === "hub") {
      feed.options.hub = href;
    }
  }
}

function buildItem(item: convert.ElementCompact, feed: Feed) {
  const date = item.pubDate?._text ? new Date(item.pubDate._text) : new Date();
  const feedItem = new FeedItem({
    id: item.guid?._text || "",
    title: {
      text:
        (item.title?._cdata as string) || (item.title?._text as string) || "",
      type: defaultTextType,
    },
    link: item.link?._text || "",
    date,
    published: date,
    description: {
      text: item.description?._cdata || item.description?._text || "",
      type: defaultTextType,
    },
    content: {
      text:
        item["content:encoded"]?._cdata || item["content:encoded"]?._text || "",
      type: defaultTextType,
    },
  });

  if (item.author) {
    if (Array.isArray(item.author)) {
      item.author.forEach((author) => {
        if (author._text) {
          feedItem.addAuthor(parseItemAuthor(author._text));
        }
      });
    } else if (item.author._text) {
      feedItem.addAuthor(parseItemAuthor(item.author._text));
    }
  }

  if (item.category) {
    if (Array.isArray(item.category)) {
      item.category.forEach((category) => {
        if (category._text) {
          feedItem.addCategory({
            name: category._text,
            domain: category._attributes?.domain,
          });
        }
      });
    } else if (item.category._text) {
      feedItem.addCategory({
        name: item.category._text,
        domain: item.category._attributes?.domain,
      });
    }
  }

  if (item.enclosure) {
    parseItemEnclosure(item.enclosure, feedItem);
  }

  Object.keys(item)
    .filter((key) => !combinedFeedItemFields.includes(key))
    .forEach((key) => {
      const value = item[key];
      if (Array.isArray(value)) {
        feedItem.setCustomField(
          key,
          value.map((item) => item?._text),
        );
      } else {
        feedItem.setCustomField(key, value?._text);
      }
    });

  feed.addItem(feedItem);
}

function parseItemAuthor(author: string): Author {
  const authorRegex = /^(.*)\((.*)\)$/;
  const match = author.match(authorRegex);
  if (match) {
    return {
      email: match[1].trim(),
      name: match[2].trim(),
    };
  }
  return {
    name: author,
    email: "",
  };
}

function parseItemEnclosure(
  enclosure: convert.ElementCompact,
  feedItem: FeedItem,
) {
  const originalType = enclosure._attributes?.type?.toString();
  let type = "image";
  if (originalType && originalType.includes("/")) {
    type = originalType.split("/")[0];
  }
  let length: number | undefined = undefined;
  if (enclosure._attributes?.length) {
    const num = parseInt(enclosure._attributes?.length.toString());
    if (!isNaN(num)) {
      length = num;
    }
  }
  const result: Enclosure = {
    url: enclosure._attributes?.url?.toString() || "",
    type: originalType,
    length,
    title: enclosure._attributes?.title?.toString(),
  };
  feedItem.options.enclosure = result;
  if (type === "image") {
    feedItem.options.image = result;
  } else if (type === "audio") {
    feedItem.options.audio = result;
  } else if (type === "video") {
    feedItem.options.video = result;
  }
}

function parseStylesheet(xml: convert.ElementCompact): string | undefined {
  if (xml._instruction && xml._instruction["xml-stylesheet"]) {
    const attrs = xml._instruction["xml-stylesheet"].split(" ");
    const hrefAttr = attrs.find((attr) => attr.startsWith("href="))?.split("=");
    if (hrefAttr && hrefAttr.length > 1 && hrefAttr[1]) {
      return hrefAttr[1].replace(/["]/g, "");
    }
  }
  return undefined;
}
