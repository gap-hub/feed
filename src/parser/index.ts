import axios from "axios";
import * as convert from "xml-js";

import { Feed } from "../feed";
import { FeedParseOptions } from "../typings";
import { isURL } from "../utils";
import { parseAtom } from "./atom";
import { parseJSON } from "./json";
import { parseRSS } from "./rss";

const DEFAULT_HEADERS = {
  "User-Agent": "@gaphub/feed-parser",
  Accept: "application/rss+xml, application/atom+xml, application/json",
};
const DEFAULT_TIMEOUT = 60_000;
const DEFAULT_MAX_REDIRECTS = 5;

const defaultFeedParseOptions: FeedParseOptions = {
  timeout: DEFAULT_TIMEOUT,
  maxRedirects: DEFAULT_MAX_REDIRECTS,
};

export enum ContentType {
  XML,
  JSON,
}

export class FeedParser {
  private readonly etags = new Map<string, string>();
  private readonly lastModified = new Map<string, string>();

  async parse(
    feedURLOrContent: string,
    options?: FeedParseOptions,
  ): Promise<Feed | null> {
    const opts = {
      ...defaultFeedParseOptions,
      ...options,
    };
    const content = await this.getContent(feedURLOrContent, opts);
    return await this.parseString(content);
  }

  async parseString(content: string): Promise<Feed | null> {
    try {
      const jsonFeed = JSON.parse(content);
      return parseJSON(jsonFeed);
    } catch {
      const xmlFeed: convert.ElementCompact = convert.xml2js(content, {
        compact: true,
      });
      if (xmlFeed.feed) {
        return parseAtom(xmlFeed);
      } else if (xmlFeed.rss) {
        return parseRSS(xmlFeed);
      }
    }
    return null;
  }

  private async getContent(
    urlOrContent: string,
    options: FeedParseOptions,
  ): Promise<string> {
    if (!isURL(urlOrContent)) {
      return urlOrContent;
    }
    return (await this.getContentFromURL(urlOrContent, options)) ?? "";
  }

  private async getContentFromURL(
    feedURL: string,
    options: FeedParseOptions,
  ): Promise<string | null> {
    const headers: Record<string, string> = {
      ...DEFAULT_HEADERS,
      ...options?.headers,
    };
    if (this.etags.has(feedURL)) {
      headers["If-None-Match"] = this.etags.get(feedURL)!;
    }
    if (this.lastModified.has(feedURL)) {
      headers["If-Modified-Since"] = this.lastModified.get(feedURL)!;
    }
    const response = await axios.get<string>(feedURL, {
      headers,
      timeout: options.timeout,
      maxRedirects: options.maxRedirects,
      responseType: "text",
    });
    if (response.status === 304) {
      // Not modified
      return null;
    } else if (response.status >= 300) {
      throw new Error(`Unexpected status code: ${response.status}`);
    }
    if (response.headers["etag"]) {
      this.etags.set(feedURL, response.headers["etag"]);
    }
    if (response.headers["last-modified"]) {
      this.lastModified.set(feedURL, response.headers["last-modified"]);
    }
    return response.data;
  }
}
