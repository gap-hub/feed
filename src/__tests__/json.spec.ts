import { FeedParser } from "../parser";
import { sampleFeed } from "./setup";

describe("json 1", () => {
  it("should generate a valid feed and parse back to feed", async () => {
    const actual = sampleFeed.json1();
    expect(actual).toMatchSnapshot("generated");

    const parser = new FeedParser();
    const parsedFeed = await parser.parseString(actual);
    expect(parsedFeed).not.toBeNull();
    expect(parsedFeed!.json1()).toMatchSnapshot("parsed");
  });
});
