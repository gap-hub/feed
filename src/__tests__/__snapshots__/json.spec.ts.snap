// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`json 1 should generate a valid feed and parse back to feed: generated 1`] = `
"{
    "version": "https://jsonfeed.org/version/1.1",
    "title": "Feed Title",
    "home_page_url": "http://example.com/",
    "feed_url": "http://example.com/sampleFeed.json",
    "description": "This is my personnal feed!",
    "icon": "http://example.com/image.png",
    "favicon": "http://example.com/image.ico",
    "language": "en",
    "authors": [
        {
            "name": "John Doe",
            "url": "https://example.com/johndoe?link=sanitized&value=2"
        }
    ],
    "_example_extension": {
        "about": "just an extension example",
        "dummy": "example"
    },
    "customField": "customValue",
    "customFieldArray": [
        "customValue1",
        "customValue2"
    ],
    "items": [
        {
            "id": "https://example.com/hello-world?id=this&that=true",
            "content_text": "Content of my item",
            "url": "https://example.com/hello-world?link=sanitized&value=2",
            "title": "Hello World",
            "summary": "This is an article about Hello World.",
            "image": "https://example.com/hello-world.jpg",
            "date_modified": "2013-07-13T23:00:00.000Z",
            "date_published": "2013-07-10T23:00:00.000Z",
            "authors": [
                {
                    "name": "Jane Doe",
                    "url": "https://example.com/janedoe?link=sanitized&value=2"
                },
                {
                    "name": "Joe Smith",
                    "url": "https://example.com/joesmith"
                },
                {
                    "name": "Joe Smith, Name Only"
                }
            ],
            "tags": [
                "Grateful Dead",
                "MSFT"
            ],
            "_item_extension_1": {
                "about": "just an item extension example",
                "dummy1": "example"
            },
            "_item_extension_2": {
                "about": "just a second item extension example",
                "dummy1": "example"
            },
            "customFieldItem": "customValueItem",
            "customFieldItemArray": [
                "customValueItem1",
                "customValueItem2"
            ]
        }
    ]
}"
`;

exports[`json 1 should generate a valid feed and parse back to feed: parsed 1`] = `
"{
    "version": "https://jsonfeed.org/version/1.1",
    "title": "Feed Title",
    "home_page_url": "http://example.com/",
    "feed_url": "http://example.com/sampleFeed.json",
    "description": "This is my personnal feed!",
    "icon": "http://example.com/image.png",
    "favicon": "http://example.com/image.ico",
    "language": "en",
    "authors": [
        {
            "name": "John Doe",
            "url": "https://example.com/johndoe?link=sanitized&value=2"
        }
    ],
    "_example_extension": {
        "about": "just an extension example",
        "dummy": "example"
    },
    "customField": "customValue",
    "customFieldArray": [
        "customValue1",
        "customValue2"
    ],
    "items": [
        {
            "id": "https://example.com/hello-world?id=this&that=true",
            "content_text": "Content of my item",
            "url": "https://example.com/hello-world?link=sanitized&value=2",
            "title": "Hello World",
            "summary": "This is an article about Hello World.",
            "image": "https://example.com/hello-world.jpg",
            "date_modified": "2013-07-13T23:00:00.000Z",
            "date_published": "2013-07-10T23:00:00.000Z",
            "authors": [
                {
                    "name": "Jane Doe",
                    "url": "https://example.com/janedoe?link=sanitized&value=2"
                },
                {
                    "name": "Joe Smith",
                    "url": "https://example.com/joesmith"
                },
                {
                    "name": "Joe Smith, Name Only"
                }
            ],
            "tags": [
                "Grateful Dead",
                "MSFT"
            ],
            "_item_extension_1": {
                "about": "just an item extension example",
                "dummy1": "example"
            },
            "_item_extension_2": {
                "about": "just a second item extension example",
                "dummy1": "example"
            },
            "customFieldItem": "customValueItem",
            "customFieldItemArray": [
                "customValueItem1",
                "customValueItem2"
            ]
        }
    ]
}"
`;
